trl/trainer/srpo_trainer.py

```diff
@@ -0,0 +1,1755 @@
+
+    def store_metrics(
+        self, metrics: Dict[str, float], train_eval: Literal["train", "eval"] = "train"
+    ) -> None:
+        for key, value in metrics.items():
+            self._stored_metrics[train_eval][key].append(value)
+
+    def evaluation_loop(
+        self,
+        dataloader: DataLoader,
+        description: str,
+        prediction_loss_only: Optional[bool] = None,
+        ignore_keys: Optional[List[str]] = None,
+        metric_key_prefix: str = "eval",
+    ) -> EvalLoopOutput:
+        """
+        Overriding built-in evaluation loop to store metrics for each batch.
+        Prediction/evaluation loop, shared by `Trainer.evaluate()` and `Trainer.predict()`.
+
+        Works both with or without labels.
+        """
+
+        # Sample and save to game log if requested (for one batch to save time)
+        if self.generate_during_eval:
+            # Generate random indices within the range of the total number of samples
+            num_samples = len(dataloader.dataset)
+            random_indices = random.sample(
+                range(num_samples), k=self.args.eval_batch_size
+            )
+
+            # Use dataloader.dataset.select to get the random batch without iterating over the DataLoader
+            random_batch_dataset = dataloader.dataset.select(random_indices)
+            random_batch = self.data_collator(random_batch_dataset)
+            random_batch = self._prepare_inputs(random_batch)
+
+            policy_output_decoded, ref_output_decoded, revisions = (
+                self.get_batch_samples(self.model, random_batch)
+            )
+
+            rows = []
+            # self.tokenizer.batch_decode(random_batch["zero_prompt_prompt_input_ids"])
+            for prompt, pol, ref, revision in zip(
+                random_batch["prompt"],
+                policy_output_decoded,
+                ref_output_decoded,
+                revisions,
+            ):
+                rows.append(
+                    [
+                        prompt,
+                        pol[len(prompt) :],
+                        ref[len(prompt) :],
+                        revision[0],
+                        revision[1],
+                        revision[2],
+                        revision[3],
+                        revision[4],
+                        revision[5],
+                    ]
+                )
+            wandb_table = {
+                "game_log": wandb.Table(
+                    columns=[
+                        "Prompt",
+                        "Policy",
+                        "Ref Model",
+                        "0 Revision",
+                        "1 Revision",
+                        "2 Revision",
+                        "3 Revision",
+                        "4 Revision",
+                        "5 Revision",
+                    ],
+                    rows=rows,
+                )
+            }
+            wandb.log(wandb_table)
```

Comment:
I was getting an error here with the existing DPO way of logging. Probably worth looking into.